internal/pkg/agent/application/upgrade/marker_watcher.go

```diff
@@ -0,0 +1,131 @@
+}
+
+func (mfw *MarkerFileWatcher) Run(ctx context.Context) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create upgrade marker watcher: %w", err)
+	}
+
+	// Watch the upgrade marker file's directory, not the file itself, so we
+	// notice the file even if it's deleted and recreated.
+	upgradeMarkerDirPath := filepath.Dir(mfw.markerFilePath)
+	if err := watcher.Add(upgradeMarkerDirPath); err != nil {
+		return fmt.Errorf("failed to set watch on upgrade marker's directory [%s]: %w", upgradeMarkerDirPath, err)
+	}
+
+	// Do an initial read from the upgrade marker file, in case the file
+	// is already present before the watching starts.
+	doInitialRead := make(chan struct{}, 1)
+	doInitialRead <- struct{}{}
+
+	// Handle watching
+	go func() {
+		defer watcher.Close()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err, ok := <-watcher.Errors:
+				if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+					mfw.logger.Debug("fsnotify.Watcher's error channel was closed")
+					return
+				}
+				mfw.logger.Errorf("upgrade marker watch returned error: %s", err)
+				continue
+			case e, ok := <-watcher.Events:
+				if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+					mfw.logger.Debug("fsnotify.Watcher's events channel was closed")
+					return
+				}
+
+				if e.Name != mfw.markerFilePath {
+					// Since we are watching the directory that will contain the upgrade
+					// marker file, we could receive events here for changes to files other
+					// than the upgrade marker. We ignore such events as we're only concerned
+					// with changes to the upgrade marker.
+					continue
+				}
+
+				switch {
+				case e.Op&(fsnotify.Create|fsnotify.Write) != 0:
+					// Upgrade marker file was created or updated; read its contents
+					// and send them over the update channel.
+					mfw.processMarker()
+				}
+			case <-doInitialRead:
+				mfw.processMarker()
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (mfw *MarkerFileWatcher) processMarker() {
+	marker, err := loadMarker(mfw.markerFilePath)
+	if err != nil {
+		mfw.logger.Error(err)
+		return
+	}
+
+	// Nothing to do if marker is not (yet) present
+	if marker == nil {
+		return
+	}
+
+	// If we are able to load (read) the upgrade marker here, but there are no
+	// upgrade details in it, it's possible (particularly on Windows, due to
+	// multiple processes reading and writing from the same file), that
+	// the Upgrade Watcher failed to update the upgrade marker file with
+	// upgrade details in it. In such a scenario, if we detect that the
+	// version of Agent we're running right now is the same as the prevVersion
+	// recorded in the upgrade marker we just read here, it would mean the
+	// upgrade was rolled back. So we explicitly create upgrade details in the
+	// upgrade marker with the UPG_ROLLBACK state.
+	if marker.Details == nil && marker.PrevVersion == version.GetAgentPackageVersion() {
```

Comment:
Does this handle the case where the watcher failed to write the rolled back state? Isn't it possible for details to not be nil, but be stuck in the watching/monitoring state in this case?

Reply:
You're right.  This sufficient condition for detecting rollback here should be if the marker is present but `marker.PrevVersion == version.GetAgentPackageVersion()`.    Fixing.

Reply:
Fixed in f735acd21d.

Reply:
With added/updated unit tests: 99883dd3dc